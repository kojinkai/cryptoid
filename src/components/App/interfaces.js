export interface Wallet {
  name: string,
  id: string,
  balance: {
    currency: string,
    amount: string
  },
  currency: string,
  purchases: string[]
}