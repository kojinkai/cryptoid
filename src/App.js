// @flow
import React, { Component } from 'react';
import './App.css';
import coinbaseApi from './api/api';
import Masthead    from './components/masthead/masthead';
import Aggregator  from './components/aggregator/aggregator';
import Tabs        from './components/tabs/tabs';

interface Account {
  name: string,
  id: string,
  balance: {
    currency: string,
    amount: string
  },
  currency: string
}

class App extends Component {

  state: {
    accounts: Array<Account>,
    activeAccount: Account,
    purchases: {},
    isLoading: boolean
  };

  switchTab: any;

  constructor(props: {}) {
    super(props);

    this.state = {
      accounts: [{
        name: '',
        id: '',
        balance: {
          currency: '',
          amount: ''
        },
        currency: ''
      }],
      activeAccount: {
        name: '',
        id: '',
        balance: {
          currency: '',
          amount: ''
        },
        currency: ''
      },    
      purchases: {},
      isLoading: true
    };

    this.switchTab = this.switchTab.bind(this);
  }

  _addOverallAccount(accountData: { data: Array<Account> }): Array<Account> {

    const { data } = accountData;
    
    const overallAccountSum = data.reduce((accumulator: number, account: Account) => {
      return accumulator + parseFloat(account.balance.amount);
    }, 0);

    const overallAccount: Account = {
      name: 'ALL',
      id: '12345',
      balance: {
        amount: overallAccountSum.toString(),
        currency: 'GBP'
      },
      currency: 'GBP'
    }

    const allAccounts: Array<Account> = [...data];

    // splice all accounts in the middle for better layout.
    allAccounts.splice(1, 0, overallAccount);
    return allAccounts;   
  }

  _getAccountByName(name: string, allAccounts: Array<Account>): Account {
    return allAccounts.filter(account => account.name === name)[0];
  }

  switchTab(name: string) {
    const activeAccount: Account = this._getAccountByName(name, this.state.accounts);
    this.setState({ activeAccount });
  }

  render() {
    return (
      <div className="App">
        <Masthead />
        <Tabs 
          active={this.state.activeAccount}
          accounts={this.state.accounts}
          switchtab={this.switchTab}
        />
        <Aggregator 
          account={this.state.activeAccount}
          purchases={this.state.purchases}
          isLoading={this.state.isLoading}
        />
      </div>
    );
  }

  componentDidMount() {

    coinbaseApi.getAccount()
      .then(response => response.json())
      .then(this._addOverallAccount)
      .then(allAccounts => this.setState({
        accounts: allAccounts,
        activeAccount: allAccounts.filter(account => account.name === 'ALL')[0]
      }))
      .then(() => coinbaseApi.getBuys())
      .then(response => response.json())
      .then(purchaseData => this.setState({ purchases: purchaseData.data, isLoading: false }))
      .catch(() => {
        console.log('api request failed');
      });


  }
}

export default App;
