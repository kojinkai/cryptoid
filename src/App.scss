@import 'styles/colors';
.App {
  text-align: center;
}
