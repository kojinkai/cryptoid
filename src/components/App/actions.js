// @flow

import { createAction } from 'redux-actions';
import coinbaseApi from '../../api/api';
import { Wallet } from './interfaces';

export const getAccountData = createAction('GET_ACCOUNT_DATA');
export const receiveAccountData = createAction('RECEIVE_ACCOUNT_DATA');
export const errorAccountData = createAction('ERROR_ACCOUNT_DATA');
export const finishLoadingAccountData = createAction('FINISH_ACCOUNT_DATA');
export const switchWallet = createAction('SWITCH_WALLET');

export const fetchAccounts = () => dispatch => {
  dispatch(getAccountData());

  return coinbaseApi.getAccounts()
    .then(response => response.json())
    .then(accountData => {     
      accountData.data.forEach((account, index) =>
        coinbaseApi.getAccountPurchasesByID(account.id)
          .then(response => response.json())
          .then(purchases => {
            const accountWithPurchases: Array<Wallet> = {...account};
            accountWithPurchases.purchases = purchases.data;
            return accountWithPurchases;
          })
          .then(accountsWithPurchases => ({
              wallet: accountsWithPurchases,
            }))
          .then(accountState => dispatch(receiveAccountData(accountState)))
          .then(() => index === accountData.data.length - 1 && dispatch(finishLoadingAccountData()))
      )
    })
    .catch(err => console.error(err));
  };
