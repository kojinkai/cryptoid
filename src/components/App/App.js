// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux'
import './App.css';
import { fetchAccounts } from './actions';
import Masthead    from '../masthead/masthead';
import Aggregator  from '../aggregator';
import Tabs        from '../tabs';

class App extends Component {

  constructor(props) {
    super(props);
    this.switchWallet = this.switchWallet.bind(this);
  }

  switchWallet: any;

  switchWallet(name: string) {
    this.props.switchWallet(name);
  }

  render() {
    return (
      <div className="App">
        <Masthead />
        <Tabs onSwitchTab={this.switchWallet} />
        <Aggregator />
      </div>
    );
  }

  componentDidMount() {
    const { dispatch } = this.props
    dispatch(fetchAccounts())
  }
}

export default connect()(App);
