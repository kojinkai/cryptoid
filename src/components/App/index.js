import { connect } from 'react-redux';
import App from './App';
import { switchWallet, fetchAccounts } from './actions';

const mapStateToProps = state => ({
  activeWallet: state.account.get('activeWallet').toJS(),
  isLoading: state.account.get('isLoading')
});

const mapDispatchToProps = dispatch => ({
  switchWallet: tabName => dispatch(switchWallet(tabName)),
  getAccountData: () => dispatch(fetchAccounts()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
