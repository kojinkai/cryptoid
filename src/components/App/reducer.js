import { fromJS } from 'immutable';
import { handleActions } from 'redux-actions';
import { getAccountData, receiveAccountData, finishLoadingAccountData } from './actions';

const defaultState = fromJS({ wallets: [], isLoading: false });

const appReducer = handleActions({
  [getAccountData]: state => state.set('isLoading', true),
  [receiveAccountData]: (state, action) => state.update('wallets', wallets => wallets.push(action.payload.wallet)),
  [finishLoadingAccountData]: state => state.set('isLoading', false),
}, defaultState);

export default appReducer;
