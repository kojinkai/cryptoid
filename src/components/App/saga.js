import { all, takeEvery, call } from 'redux-saga/effects'
// import { getAccountData, receiveAccountData, finishLoadingAccountData } from './actions';
import { getAccountData } from './actions';
import coinbaseApi from '../../api/api';

export function* getAccounts() {
  const products = yield call(coinbaseApi.getAccounts);
  console.log(products);
}

function* watchGetAccountData() {
  yield takeEvery(getAccountData().type, getAccounts);
}

export default function* rootSaga() {
  yield all([
    watchGetAccountData(),
    getAccounts(),
  ])
}