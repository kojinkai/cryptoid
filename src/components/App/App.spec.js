import { componentFactory } from '../../test/helpers';
import wallets from '../../test/mocks/wallets';
import App from './App';

const createComponentWithProps = componentFactory(App, {
  switchWallet: jest.fn(),
  getAccountData: jest.fn(),
  activeWallet: wallets[0],
  isLoading: false,
});

describe('<App />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = createComponentWithProps();
  });

  it('should render according the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call for the account data on mount', () => {
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().props.getAccountData).toHaveBeenCalled();
  });

  it('should call through to the switchWallet prop when the switchWallet method is called', () => {
    wrapper.instance().switchWallet();
    expect(wrapper.instance().props.getAccountData).toHaveBeenCalled();
  });
});
