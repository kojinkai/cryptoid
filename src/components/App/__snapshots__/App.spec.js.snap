// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render according the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="App"
>
    <Masthead />
    <Connect(Tabs)
        onSwitchTab={[Function]}
    />
    <Aggregator
        activeWallet={
            Object {
                "balance": Object {
                  "amount": "800",
                  "currency": "GBP",
                },
                "currency": "BTC",
                "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
                "name": "BTC",
                "purchases": Array [
                  Object {
                    "currency": "BTC",
                    "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                    "total": Object {
                      "amount": "101.01",
                      "currency": "GBP",
                    },
                  },
                  Object {
                    "currency": "BTC",
                    "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                    "total": Object {
                      "amount": "202.02",
                      "currency": "GBP",
                    },
                  },
                  Object {
                    "currency": "BTC",
                    "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                    "total": Object {
                      "amount": "303.03",
                      "currency": "GBP",
                    },
                  },
                ],
              }
        }
        isLoading={false}
    />
    <Purchases
        activeWalletName="BTC"
        isLoading={false}
        purchases={
            Array [
                Object {
                  "currency": "BTC",
                  "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                  "total": Object {
                    "amount": "101.01",
                    "currency": "GBP",
                  },
                },
                Object {
                  "currency": "BTC",
                  "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                  "total": Object {
                    "amount": "202.02",
                    "currency": "GBP",
                  },
                },
                Object {
                  "currency": "BTC",
                  "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                  "total": Object {
                    "amount": "303.03",
                    "currency": "GBP",
                  },
                },
              ]
        }
    />
</div>,
  "nodes": Array [
    <div
      className="App"
>
      <Masthead />
      <Connect(Tabs)
            onSwitchTab={[Function]}
      />
      <Aggregator
            activeWallet={
                  Object {
                        "balance": Object {
                          "amount": "800",
                          "currency": "GBP",
                        },
                        "currency": "BTC",
                        "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
                        "name": "BTC",
                        "purchases": Array [
                          Object {
                            "currency": "BTC",
                            "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "101.01",
                              "currency": "GBP",
                            },
                          },
                          Object {
                            "currency": "BTC",
                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "202.02",
                              "currency": "GBP",
                            },
                          },
                          Object {
                            "currency": "BTC",
                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "303.03",
                              "currency": "GBP",
                            },
                          },
                        ],
                      }
            }
            isLoading={false}
      />
      <Purchases
            activeWalletName="BTC"
            isLoading={false}
            purchases={
                  Array [
                        Object {
                          "currency": "BTC",
                          "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                          "total": Object {
                            "amount": "101.01",
                            "currency": "GBP",
                          },
                        },
                        Object {
                          "currency": "BTC",
                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                          "total": Object {
                            "amount": "202.02",
                            "currency": "GBP",
                          },
                        },
                        Object {
                          "currency": "BTC",
                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                          "total": Object {
                            "amount": "303.03",
                            "currency": "GBP",
                          },
                        },
                      ]
            }
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <App
        activeWallet={
                Object {
                        "balance": Object {
                          "amount": "800",
                          "currency": "GBP",
                        },
                        "currency": "BTC",
                        "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
                        "name": "BTC",
                        "purchases": Array [
                          Object {
                            "currency": "BTC",
                            "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "101.01",
                              "currency": "GBP",
                            },
                          },
                          Object {
                            "currency": "BTC",
                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "202.02",
                              "currency": "GBP",
                            },
                          },
                          Object {
                            "currency": "BTC",
                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                            "total": Object {
                              "amount": "303.03",
                              "currency": "GBP",
                            },
                          },
                        ],
                      }
        }
        getAccountData={[Function]}
        isLoading={false}
        switchWallet={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": App {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "activeWallet": Object {
            "balance": Object {
              "amount": "800",
              "currency": "GBP",
            },
            "currency": "BTC",
            "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
            "name": "BTC",
            "purchases": Array [
              Object {
                "currency": "BTC",
                "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "101.01",
                  "currency": "GBP",
                },
              },
              Object {
                "currency": "BTC",
                "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "202.02",
                  "currency": "GBP",
                },
              },
              Object {
                "currency": "BTC",
                "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "303.03",
                  "currency": "GBP",
                },
              },
            ],
          },
          "getAccountData": [Function],
          "isLoading": false,
          "switchWallet": [Function],
        },
        "refs": Object {},
        "state": null,
        "switchWallet": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="App"
>
          <Masthead />
          <Connect(Tabs)
                    onSwitchTab={[Function]}
          />
          <Aggregator
                    activeWallet={
                              Object {
                                        "balance": Object {
                                          "amount": "800",
                                          "currency": "GBP",
                                        },
                                        "currency": "BTC",
                                        "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
                                        "name": "BTC",
                                        "purchases": Array [
                                          Object {
                                            "currency": "BTC",
                                            "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "101.01",
                                              "currency": "GBP",
                                            },
                                          },
                                          Object {
                                            "currency": "BTC",
                                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "202.02",
                                              "currency": "GBP",
                                            },
                                          },
                                          Object {
                                            "currency": "BTC",
                                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "303.03",
                                              "currency": "GBP",
                                            },
                                          },
                                        ],
                                      }
                    }
                    isLoading={false}
          />
          <Purchases
                    activeWalletName="BTC"
                    isLoading={false}
                    purchases={
                              Array [
                                        Object {
                                          "currency": "BTC",
                                          "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "101.01",
                                            "currency": "GBP",
                                          },
                                        },
                                        Object {
                                          "currency": "BTC",
                                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "202.02",
                                            "currency": "GBP",
                                          },
                                        },
                                        Object {
                                          "currency": "BTC",
                                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "303.03",
                                            "currency": "GBP",
                                          },
                                        },
                                      ]
                    }
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="App"
>
          <Masthead />
          <Connect(Tabs)
                    onSwitchTab={[Function]}
          />
          <Aggregator
                    activeWallet={
                              Object {
                                        "balance": Object {
                                          "amount": "800",
                                          "currency": "GBP",
                                        },
                                        "currency": "BTC",
                                        "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
                                        "name": "BTC",
                                        "purchases": Array [
                                          Object {
                                            "currency": "BTC",
                                            "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "101.01",
                                              "currency": "GBP",
                                            },
                                          },
                                          Object {
                                            "currency": "BTC",
                                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "202.02",
                                              "currency": "GBP",
                                            },
                                          },
                                          Object {
                                            "currency": "BTC",
                                            "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                            "total": Object {
                                              "amount": "303.03",
                                              "currency": "GBP",
                                            },
                                          },
                                        ],
                                      }
                    }
                    isLoading={false}
          />
          <Purchases
                    activeWalletName="BTC"
                    isLoading={false}
                    purchases={
                              Array [
                                        Object {
                                          "currency": "BTC",
                                          "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "101.01",
                                            "currency": "GBP",
                                          },
                                        },
                                        Object {
                                          "currency": "BTC",
                                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "202.02",
                                            "currency": "GBP",
                                          },
                                        },
                                        Object {
                                          "currency": "BTC",
                                          "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                                          "total": Object {
                                            "amount": "303.03",
                                            "currency": "GBP",
                                          },
                                        },
                                      ]
                    }
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <App
    activeWallet={
        Object {
            "balance": Object {
              "amount": "800",
              "currency": "GBP",
            },
            "currency": "BTC",
            "id": "58542935-67b5-56e1-a3f9-42686e07fa40",
            "name": "BTC",
            "purchases": Array [
              Object {
                "currency": "BTC",
                "id": "9e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "101.01",
                  "currency": "GBP",
                },
              },
              Object {
                "currency": "BTC",
                "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "202.02",
                  "currency": "GBP",
                },
              },
              Object {
                "currency": "BTC",
                "id": "8e14d574-30fa-5d85-b02c-6be0d851d61d",
                "total": Object {
                  "amount": "303.03",
                  "currency": "GBP",
                },
              },
            ],
          }
    }
    getAccountData={[Function]}
    isLoading={false}
    switchWallet={[Function]}
/>,
}
`;
